### PR TITLE
enable etcdv3 client logging

### LINKED ISSUE #64730: Provide a way to enable etcdv3 client logging.
/kind feature

Can we provide a way to enable etcd clientv3 logging at the apiserver side ? 
The etcd.clientv3 interface is instantiated [here](https://github.com/ahakanbaba/kubernetes/blob/4f1489edf0675718bb4a5ac754677e2450215d17/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go#L44) and it does not set a Logger. 

Without a logger object, etcd.clientv3 object discards the [logs](https://github.com/ahakanbaba/kubernetes/blob/4f1489edf0675718bb4a5ac754677e2450215d17/vendor/github.com/coreos/etcd/clientv3/client.go#L275)

It would be nice to have a command line argument, config/environment variable to enable logging in the etcd clientv3 interface. 

Could the glog functions be wrapped inside an object that implements the grpclog.Logger interface.

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
provide a way to enable etcd clientv3 logging at the apiserver side.

This pr is almost copied from http://paste.openstack.org/show/722687/

fixes: #64730

**Release note**:

```release-note
Enable etcdv3 client logging.
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go [hunk 1]
```diff
@@ -79,3 +87,60 @@ func newETCD3Storage(c storagebackend.Config) (storage.Interface, DestroyFunc, e
 	}
 	return etcd3.NewWithNoQuorumRead(client, c.Codec, c.Prefix, transformer, c.Paging), destroyFunc, nil
 }
+
+// loggerT is the default logger that delegates to glog and used by grpclog
+type loggerT struct{}
+
+var _ clientv3.Logger = &loggerT{}
```
Comment (ahakanbaba) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: Why is this line necessary ?

Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: to make sure `loggerT`  implements `clientv3.Logger`. If we missed one method, it will fail in compilation.

**Comment by hzxuzhonghu** [comment]
cc @dims @xiang90 @liggitt  ptal

**Comment by dims** [comment]
@hzxuzhonghu if you take a typical kubeadm deployment and try this patch out, how many more lines do you see in api server logs? (compare before and after). So we get an idea if we have to dial the logging a bit down ...

**Comment by hzxuzhonghu** [comment]
@dims After this ,  logs print during startup:

```
I0621 09:47:02.786329   68206 etcd3.go:105] ccBalancerWrapper: updating state and picker called by balancer: IDLE, 0xc42a8503c0
I0621 09:47:02.786374   68206 etcd3.go:105] dialing to target with scheme: ""
I0621 09:47:02.786392   68206 etcd3.go:105] could not get resolver for scheme: ""
I0621 09:47:02.786420   68206 etcd3.go:105] balancerWrapper: is pickfirst: false
I0621 09:47:02.786428   68206 etcd3.go:105] balancerWrapper: got update addr from Notify: [{127.0.0.1:2379 <nil>}]
I0621 09:47:02.786444   68206 etcd3.go:105] ccBalancerWrapper: new subconn: [{127.0.0.1:2379 0  <nil>}]
I0621 09:47:02.786560   68206 etcd3.go:105] balancerWrapper: handle subconn state change: 0xc42a7fa890, CONNECTING
I0621 09:47:02.789154   68206 etcd3.go:105] ccBalancerWrapper: updating state and picker called by balancer: CONNECTING, 0xc42a8503c0
I0621 09:47:02.789174   68206 etcd3.go:105] balancerWrapper: handle subconn state change: 0xc42a7fa890, READY
I0621 09:47:02.789196   68206 etcd3.go:105] ccBalancerWrapper: updating state and picker called by balancer: READY, 0xc42a8503c0
I0621 09:47:02.789239   68206 etcd3.go:105] balancerWrapper: got update addr from Notify: [{127.0.0.1:2379 <nil>}]
I0621 09:47:02.789650   68206 store.go:1397] Monitoring apiservices.apiregistration.k8s.io count at <storage-prefix>//apiregistration.k8s.io/apiservices
I0621 09:47:02.789690   68206 reflector.go:240] Listing and watching *apiregistration.APIService from storage/cacher.go:/apiregistration.k8s.io/apiservices
I0621 09:47:02.789703   68206 etcd3.go:105] ccBalancerWrapper: updating state and picker called by balancer: IDLE, 0xc42a7f07e0
I0621 09:47:02.789715   68206 etcd3.go:105] dialing to target with scheme: ""
I0621 09:47:02.789720   68206 etcd3.go:105] could not get resolver for scheme: ""
I0621 09:47:02.789823   68206 etcd3.go:105] balancerWrapper: is pickfirst: false
I0621 09:47:02.789831   68206 etcd3.go:105] balancerWrapper: got update addr from Notify: [{127.0.0.1:2379 <nil>}]
I0621 09:47:02.789841   68206 etcd3.go:105] ccBalancerWrapper: new subconn: [{127.0.0.1:2379 0  <nil>}]
I0621 09:47:02.789999   68206 etcd3.go:105] balancerWrapper: handle subconn state change: 0xc42a7e4980, CONNECTING
I0621 09:47:02.790018   68206 etcd3.go:105] ccBalancerWrapper: updating state and picker called by balancer: CONNECTING, 0xc42a7f07e0
I0621 09:47:02.790213   68206 etcd3.go:105] balancerWrapper: handle subconn state change: 0xc42a7e4980, READY
I0621 09:47:02.790240   68206 etcd3.go:105] ccBalancerWrapper: updating state and picker called by balancer: READY, 0xc42a7f07e0
I0621 09:47:02.790257   68206 etcd3.go:105] balancerWrapper: got update addr from Notify: [{127.0.0.1:2379 <nil>}]

```

**Comment by ahakanbaba** [comment]
I see that especially the balancerWrapper logs would be very useful to kubernetes operators.

**Comment by xiang90** [comment]
@hzxuzhonghu 

Can we make the logging less verbose by default? Only warning and error level logging should be printed.

**Comment by hzxuzhonghu** [comment]
>Can we make the logging less verbose by default? Only warning and error level logging should be printed.

sure

**Comment by hzxuzhonghu** [comment]
@xiang90 addressed

**Comment by hzxuzhonghu** [comment]
@xiang90 @liggitt ptal

**Comment by hzxuzhonghu** [comment]
rebased @xiang90 @liggitt

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go [hunk 2]
```diff
@@ -34,6 +35,13 @@ import (
 	"k8s.io/apiserver/pkg/storage/value"
 )
 
+func init() {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: *really* not a fan of doing this in an init block... can we do something in a once block when we construct etcd3 clients?

**Comment by liggitt** [comment]
@kubernetes/sig-api-machinery-pr-reviews

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go (hunk 2 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: I will look into it.

Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: Ok, Then I have to use sync once in `newETCD3Client`

**Comment by hzxuzhonghu** [comment]
@liggitt Seems impossible with sync.Once, because the test case also starts a fake etcd server, which also uses grpclog. So we can not call `clientv3.SetLogger(new(loggerT))` while others using it. 

So to guarantee the test cases,  have to put this in init phase.

**Comment by liggitt** [comment]
> So to guarantee the test cases, have to put this in init phase.

I think it is very important to avoid setup like this in init() blocks that cannot be opted out of... can you move this to its own package and call it from the kube-apiserver startup path instead?
